### PR TITLE
Fix accuracy of np.arange and np.linspace

### LINKED ISSUE #6768: numpy.arange and numpy.linspace not accurate in numba
## Reporting a bug

- [x] I have tried using the latest released version of Numba (using `numba=0.52.0=py38h51da96c_0` from conda-forge which is the latest available version, also tested with `0.51.2` from the defaults channel).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python test_numba.py'.

I am getting different results in my code depending on if I run with numba or not. The problem is that the code that numba compiles is extremely simple (calling another compiled numba function in a for loop) but still one of the values comes out different. I have spent some time to create a small reproducible code which is attached below.

[test_numba.zip](https://github.com/numba/numba/files/6049357/test_numba.zip) which contains:
```
- test_numba.py
- env.yml # to reproduce the conda environment though 
#other versions of numba (including the one from the defaults channel) fail too

# 3 numpy arrays inside hdf5 files
- nonsep-epc.hdf5
- nonsep-negatives.hdf5
- nonsep-positives.hdf5

```

The related part of the code is (the `epc` function results are different from `epc.py_func`):
```python
@jit(nopython=True)
def min_weighted_error_rate_threshold(negatives, positives, cost):
    if cost > 1.0:
        cost = 1.0
    elif cost < 0.0:
        cost = 0.0

    return _minimizing_threshold(negatives, positives, cost)


@jit(nopython=True)
def epc(
    dev_negatives,
    dev_positives,
    n_points,
    is_sorted=False,
):
    # if not pre-sorted, copies and sorts
    dev_neg = dev_negatives if is_sorted else numpy.sort(dev_negatives)
    dev_pos = dev_positives if is_sorted else numpy.sort(dev_positives)
    step = 1.0 / (n_points - 1.0)
    alpha = numpy.arange(0, 1 + step, step)
   # something goes wrong from here
    thres = numpy.empty_like(alpha)
    for i, cost in enumerate(alpha):
        thres[i] = min_weighted_error_rate_threshold(dev_neg, dev_pos, cost)

    return thres, alpha
```
with this test function below written to test the `epc` function.
```python
def test():
    def _load(path):
        with h5py.File(path, "r") as f:
            return numpy.array(f["array"])

    positives = _load("nonsep-positives.hdf5")
    negatives = _load("nonsep-negatives.hdf5")
    thresholds_ref = _load("nonsep-epc.hdf5")[2]
    n_points = 100

    dev_negatives = negatives[: (negatives.shape[0] // 2)]
    dev_positives = positives[: (positives.shape[0] // 2)]

    print("Comparing pure Python to reference data.")
    thres, alpha_ref = epc.py_func(
        dev_negatives,
        dev_positives,
        n_points,
        is_sorted=False,
    )
    numpy.testing.assert_allclose(thres, thresholds_ref)
    print("All values are similar")

    print("Comparing sorted in Python to reference data.")
    thres, alpha = epc(
        numpy.sort(dev_negatives),
        numpy.sort(dev_positives),
        n_points,
        is_sorted=True,
    )
    numpy.testing.assert_allclose(alpha, alpha_ref)
    numpy.testing.assert_allclose(thres, thresholds_ref) # code fails here, this is line 181
    print("All values are similar")

    print("Comparing numba variant to reference data.")
    thres, alpha = epc(
        dev_negatives,
        dev_positives,
        n_points,
        is_sorted=False,
    )
    numpy.testing.assert_allclose(alpha, alpha_ref)
    numpy.testing.assert_allclose(thres, thresholds_ref) # it would fail here too if you comment the failing assertion above
    print("All values are similar")
```
If you run `python test_numba.py`, you will get:
```
$ python test_numba.py
Comparing pure Python to reference data.
All values are similar
Comparing sorted in Python to reference data.
Traceback (most recent call last):
  File "test_numba.py", line 197, in <module>
    test()
  File "test_numba.py", line 181, in test
    numpy.testing.assert_allclose(thres, thresholds_ref)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 1 / 100 (1%)
Max absolute difference: 0.4
Max relative difference: 0.06837607
```

If I call `min_weighted_error_rate_threshold` (which is a compiled function) from Python, everything works. But, when it is called from the `epc` function, one of the values turns out to be wrong. I have checked all the computations inside the epc function:
```python
@jit(nopython=True)
def epc(
    dev_negatives,
    dev_positives,
    n_points,
    is_sorted=False,
):
    # if not pre-sorted, copies and sorts
    dev_neg = dev_negatives if is_sorted else numpy.sort(dev_negatives)
    dev_pos = dev_positives if is_sorted else numpy.sort(dev_positives)
    step = 1.0 / (n_points - 1.0)
    alpha = numpy.arange(0, 1 + step, step)
    thres = numpy.empty_like(alpha)
    for i, cost in enumerate(alpha):
        thres[i] = min_weighted_error_rate_threshold(dev_neg, dev_pos, cost)

    return thres, alpha
```
and looks like all the values are correct (according to my tests in the `test()` function) till the code reaches:
```python
    thres = numpy.empty_like(alpha)
    for i, cost in enumerate(alpha):
        thres[i] = min_weighted_error_rate_threshold(dev_neg, dev_pos, cost)
```
Here, numba is just calling another already compiled function, so I don't understand what could go wrong here when numba compiles it. Please let me know if I can provide more information.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title, this adjusts the implementations so as to not accumulate
error and to better replicate what NumPy does.

Fixes #6768

### CODE REVIEW

File: numba/np/arrayobj.py [hunk 1]
```diff
@@ -3929,15 +3928,19 @@ def linspace(start, stop):
...
         if num == 0:
             return arr
         div = num - 1
-        delta = stop - start
-        arr[0] = start
-        for i in range(1, num):
-            arr[i] = start + delta * (i / div)
+        if div > 0:
+            delta = stop - start
+            step = np.divide(delta, div)
```
Comment (sklam) [inline numba/np/arrayobj.py:3939]: Is `np.divide` used to make sure numpy error model here?

Comment (stuartarchibald) [reply numba/np/arrayobj.py:3939]: Yes, but it's probably not necessary as the predicate would likely propagate and the div-by-zero branch would be DCE. Am happy to change.

Comment (sklam) [reply numba/np/arrayobj.py:3939]: i'd prefer the using `/`

Comment (stuartarchibald) [reply numba/np/arrayobj.py]: Done in 947c744

**Review by sklam — APPROVED** [review]
Thanks for the patch.

**Pull request merged**